compiler/z/codegen/OMRCodeGenerator.cpp

```diff
@@ -3211,6 +3211,101 @@ TR_S390Peephole::AGIReduction()
 
    return performed;
    }
+
+/* For runs with -Xgc:concurrentScavenge on hardware that doesn't support guarded storage facility,
+ * swap LGG/LLGFSG with regular loads and software read barrier sequence that checks if concurrent scavange is in progress
+ * and if the current object pointer is in the evacuate space then calls the GC helper to update the object pointer.
+ */
+
+bool
+TR_S390Peephole::SoftwareReadBarrier()
+   {
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+
+   if (TR::Compiler->target.cpu.getS390SupportsGuardedStorageFacility())
+      {
+      return false;
+      }
+
+   auto* concurrentScavangeNotActiveLabel = generateLabelSymbol(_cg);
+   TR::S390RXInstruction *load = (TR::S390RXInstruction*) _cursor;
+   TR::MemoryReference *loadMemRef = generateS390MemoryReference(*load->getMemoryReference(), 0, _cg);
+   TR::Register *loadTargetReg = _cursor->getRegisterOperand(1);
+   TR::Register *vmReg = _cg->getLinkage()->getMethodMetaDataRealRegister();
+   TR::Register *r14Reg = _cg->machine()->getS390RealRegister(_cg->getReturnAddressRegister());
+   TR::Instruction* prev = load->getPrev();
+
+   // If guarded load target and mem ref registers are the same,
+   // preserve the register before overwriting it, since we need to repeat the load after calling the GC helper.
+   bool shouldPreserveLoadReg = (loadMemRef->getBaseRegister() == loadTargetReg);
+   TR::MemoryReference *gsIntermediateAddrMemRef = generateS390MemoryReference(vmReg, offsetof(J9VMThread, gsParameters.intermediateResult), _cg);
```

Comment:
This code is very OpenJ9-specific.  I doubt it will even build on non-OpenJ9 runtimes if `OMR_GC_CONCURRENT_SCAVENGER` is defined.  Can you rework this code to isolate the OpenJ9-specific parts to that project and out of OMR?

Reply:
I added a few helpers under OMRVMEnv.hpp to get around needing offsets from J9VMThread for concurrent scavenge logic. Please let me know if that's not the way we want to go about it.